Implement manual integration with special functions

normalhuman opened the pull request:
#### References to other Issues or PRs

Fixes #9462. Fixes #15431. Fixes #15451

#### Brief description of what is fixed or changed

Several non-elementary antiderivatives are now recognized by `manualintegrate`. Additionally, cyclic integration by parts is corrected: previously it could integrate `exp(x)*sin(x)` but not `exp(x)*sin(2*x)` because additional constant factors were not handled.
The heuristics for integration by parts are slightly expanded in terms of the choices of `dv` term, now that more kinds of `dv` can be handled. For example, `x*exp(x)*sin(2*x)` uses `dv = exp(x)*sin(2*x)` which was previously not allowed.

These changes also expand the capability of `integrate`, which was previously unable to integrate, for example, `sin(x**2 + 1)` or `sqrt(4 - sin(x)**2)`

### Other comments

The rule `special_function_rule` is added first to the list in `integral_steps` because it is a quick exit, and because applying elementary calculus rules to such integrals is veritable waste of time.

This PR implements "easy parts" of https://github.com/sympy/sympy/issues/15429#issuecomment-433559265 and of https://github.com/sympy/sympy/issues/15471#issuecomment-438860606
Completely fixing #15471 will also require a change of `.subs(log(x), u)` logic, so that it transforms `log(x)*cos(log(x))/x**(S(3)/4)` into `u*cos(u)*exp(3*u/4)`. This is left for future work.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* integrals
  * Added standard non-elementary antiderivatives to manualintegrate.
<!-- END RELEASE NOTES -->

normalhuman committed:
Implement manual integration with special functions

Several non-elementary antiderivatives are now recognized by
`manualintegrate`. Additionally, cyclic integration by parts is
corrected: previously it could integrate `exp(x)*sin(x)` but not
`exp(x)*sin(2*x)` because newly created constant factors were not handled.
The heuristics for integration by parts are slightly expanded in terms
of the choices of `dv` term, now that more kinds of `dv` can be handled.
For example, `x*exp(x)*sin(2*x)` uses `dv = exp(x)*sin(2*x)` which was
previously not allowed.

jksuom commented on sympy/integrals/tests/test_integrals.py:
```diff
@@ -1435,3 +1435,8 @@ def test_issue_15457():
+
+def test_issue_15431():
+    assert integrate(x*exp(x)*log(x), x) == \
+        (x*exp(x) - exp(x))*log(x) - exp(x) + Ei(x)
```
It seems that the result is not unique. (Probably depends on cache state.)
```
Python 2.7.6 (default, Nov 13 2018, 12:45:42) 
[GCC 4.8.4] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import *
>>> from sympy.integrals.manualintegrate import manualintegrate
>>> x = symbols('x')
>>> manualintegrate(x*exp(x)*log(x), x)
(x*exp(x) - exp(x))*log(x) - exp(x) - expint(1, -x)
>>> manualintegrate(x*exp(x)*log(x), x)
(x*exp(x) - exp(x))*log(x) - exp(x) - expint(1, -x)
>>> manualintegrate(x*exp(x)*log(x), x)
(x*exp(x) - exp(x))*log(x) - exp(x) + Ei(x)
>>> manualintegrate(x*exp(x)*log(x), x)
(x*exp(x) - exp(x))*log(x) - exp(x) - expint(1, -x)
>>> manualintegrate(x*exp(x)*log(x), x)
(x*exp(x) - exp(x))*log(x) - exp(x) + Ei(x)
>>> manualintegrate(x*exp(x)*log(x), x)
(x*exp(x) - exp(x))*log(x) - exp(x) - expint(1, -x)
>>> manualintegrate(x*exp(x)*log(x), x)
(x*exp(x) - exp(x))*log(x) - exp(x) - expint(1, -x)
```

normalhuman committed:
Use a tuple instead of a dictionary for special function rules

To ensure consistent output, rules should be applied in a consistent
order. (In some integrals, more than one rule can be used, producing
equivalent but visually different results.)

jksuom commented on sympy/integrals/tests/test_manual.py:
```diff
@@ -323,12 +363,24 @@ def test_issue_2850():
-                           - 2*exp(x)*cos(2*x) - 4*Integral(exp(x)*sin(2*x), x)
+    assert manualintegrate(sin(2*x)*exp(x), x) == exp(x)*sin(2*x)/5 - 2*exp(x)*cos(2*x)/5
+    assert not contains_dont_know(integral_steps(sin(2*x)*exp(x), x)
+)
```
This probably should not be on a line of its own.

jksuom commented:
This looks fine.

normalhuman committed:
relocate a rogue close paren
